### DISCUSSION TITLE
Prevent rewrite same url twice by middleware

### DISCUSSION THREAD

**Discussion by andriiradkevych** [vercel/next.js#38463] [Help]
**middleware.ts**

```
import { NextResponse, type NextRequest } from 'next/server';

const isMobile = (userAgent: string) =>
  /iPhone|iPad|iPod|Android/i.test(userAgent);

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  const userAgent = req.headers.get('user-agent');

  console.log(pathname, 'pn', req.nextUrl);

  if (userAgent && !pathname.includes('favicon.ico')) {
    if (isMobile(userAgent)) {
      return NextResponse.rewrite(new URL(`/mobile${pathname}`, req.url));
    } else {
      return NextResponse.rewrite(new URL(`/desktop${pathname}`, req.url));
    }
  }
}

export const config = {
  matcher: ['/', '/posts/:slug'],
};

```

I've created `middleware` to make `rewrites` depend on user agent , and direct him to `mobile` or `desktop` version. My `next` version is `12.2.0` . I've used `config` , for detectiong on which routes I want to call my middleware . Problem here that when I go to `/` , middleware called twice , and when I navigate to `/posts/third-page` => `/posts/:slug` , middleware called twice

**Comment by icyJoseph** [comment]
Hi,

Is it called twice on the same `pathname`? 

```ts
  if (userAgent && !pathname.includes('favicon.ico')) {
    if (isMobile(userAgent)) {
      req.nextUrl.pathname = `/mobile${pathname}`;
      return NextResponse.rewrite(req.nextUrl);
    } else {
      req.nextUrl.pathname = `/desktop${pathname}`;
      return NextResponse.rewrite(req.nextUrl)
    }
  }
```

Also could you try out with mutating the `nextUrl` object, it shouldn't matter? I tested and it runs only once, even with your code.

**Reply by andriiradkevych** [reply]
@icyJoseph I think it happens , when on page you have `Link` with same url as you page on . It is the case for example , when you have nav menu , I've tried to set `prefetch={false}` , but as I remember it wasn't working

**Reply by icyJoseph** [reply]
You know what, I am now able to trigger this reliably.

A simple page such as:

```tsx
import type { NextPage } from "next";

const Home: NextPage = () => {
  return (
    <div>
      Home <div>more stuff</div>
    </div>
  );
};

export default Home;
```

Makes a request for `json` data, `index.json` for example, which didn't happen in `next@12.1.6` - I've observed it in `next@12.2.0` and `next.12.2.2`.

Of course, since there's no JSON data, you can see in the DevTools that the response is HTML. I wonder if this is a bug? It sure looks like it.

<img width="678" alt="Screenshot 2022-07-14 at 10 03 30" src="https://user-images.githubusercontent.com/21013447/178933095-06f0f477-18fc-4cf2-9598-8f17f7dbdd10.png">

Notice the content type. I'll dig a bit more into this. Good stuff.

**Reply by andriiradkevych** [reply]
yes , this is why I've  just stopped on version `12.1.2` , where it seems like it works ( middleware ), on 12.2 and more , I wasn't able to work with middleware for my case described above .. seems like it is a bug , or it is a feature )
